docs: add Charmcraft 4.1 release notes (#2483)
canonical/charmcraft: docs/release-notes/charmcraft-4.1.rst

.. _release-4.1:

Charmcraft 4.1 release notes
============================

*Unreleased*

Learn about the new features, changes, and fixes introduced in Charmcraft 4.1.

Requirements and compatibility
------------------------------

For development and testing, Charmcraft requires a host with a minimum of 4GB RAM
running a Linux distribution compatible with systemd.

All versions of Charmcraft require the following software:

- systemd
- `snapd`_
- Either `LXD`_ or `Multipass`_

We recommend you install the `Charmcraft snap <https://snapcraft.io/charmcraft>`_. It
comes bundled with all its dependencies.

Non-snap installations of Charmcraft have the following dependencies:

- Python 3.10 or higher
- libgit2 1.7
- `skopeo`_
- `Spread`_

What's new
----------

Charmcraft 4.1 brings the following new features.

> Support for Ubuntu 25.10 and Ubuntu 26.04 LTS bases

Reviewer: ```suggestion
Support for Ubuntu 25.10 and Ubuntu 26.04 LTS bases
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Charmcraft 4.1 adds experimental support for Ubuntu 25.10 and Ubuntu 26.04 LTS bases. As

> an interim base, the purpose of Ubuntu 25.10-based charms is for testing before the

Reviewer: > testing

I still think it would be better to say what we're testing _for_.

.. _release-4.1:

Charmcraft 4.1 release notes
============================

8 December 2025

Learn about the new features, changes, and fixes introduced in Charmcraft 4.1.

Requirements and compatibility
------------------------------

For development and testing, Charmcraft requires a host with a minimum of 4GB RAM
running a Linux distribution compatible with systemd.

All versions of Charmcraft require the following software:

- systemd
- `snapd`_
- Either `LXD`_ or `Multipass`_

We recommend you install the `Charmcraft snap <https://snapcraft.io/charmcraft>`_. It
comes bundled with all its dependencies.

Non-snap installations of Charmcraft have the following dependencies:

- Python 3.10 or higher
- libgit2 1.7
- `skopeo`_
- `Spread`_

What's new
----------

Charmcraft 4.1 brings the following new features.

Support for Ubuntu 25.10 and Ubuntu 26.04 LTS bases
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Charmcraft 4.1 adds experimental support for Ubuntu 25.10 and Ubuntu 26.04 LTS bases. As

> an interim base, the purpose of charms built on Ubuntu 25.10 is to test changes

Reviewer 2: ```suggestion
Charmcraft 4.1 adds experimental support for the Ubuntu 25.10 and Ubuntu 26.04 LTS bases.
The purpose of charms built on the Ubuntu 25.10 interim base is to test changes
```
The current sentence is structured as though "the purpose of charms" is the interim base.

.. code-block:: yaml

    requires:
      http-proxy:
        interface: http_proxy
        optional: true
        limit: 1

The init profiles for 12-factor apps now include this relation.

Support for Spring Boot profiles
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Charms that use the :ref:`spring-boot-framework-extension` can now configure active
`Spring Boot profiles
<https://docs.spring.io/spring-boot/reference/features/profiles.html>`_ at runtime.

Known issues
------------

The following issues were reported and are scheduled to be fixed in upcoming
patch releases.

See individual issue links for any mitigations.

- `#2078 <https://github.com/canonical/charmcraft/issues/2078>`_
  ``charmcraft clean`` does not clean all platforms for a charm.
- `#1990 <https://github.com/canonical/charmcraft/issues/1990>`_ Cannot stage
  packages with Charmcraft
- `#2493 <https://github.com/canonical/charmcraft/issues/2493>`_ Packing fails when
  using a separate metadata.yaml file and the ``--project-dir`` argument.
- `#2492 <https://github.com/canonical/charmcraft/issues/2492>`_ Internal error when
  uploading a charm outside of a project directory.

Fixed bugs and issues

> ---------------------

Reviewer 2: ```suggestion
```
Probably better to add this in a patch release than have an empty section until then.

> 

Reviewer 2: ```suggestion
```
Empty section

:orphan:
.. _release-4.1:

Charmcraft 4.1 release notes
============================

*Unreleased*

Learn about the new features, changes, and fixes introduced in Charmcraft 4.1.

Requirements and compatibility
------------------------------

For development and testing, Charmcraft requires a host with a minimum of 4GB RAM
running a Linux distribution compatible with systemd.

All versions of Charmcraft require the following software:

- systemd
- `snapd`_
- Either `LXD`_ or `Multipass`_

We recommend you install the `Charmcraft snap <https://snapcraft.io/charmcraft>`_. It
comes bundled with all its dependencies.

Non-snap installations of Charmcraft have the following dependencies:

- Python 3.10 or higher
- libgit2 1.7
- `skopeo`_
- `Spread`_

What's new
----------

Charmcraft 4.1 brings the following new features.

Ubuntu 25.10 and 26.04

> ~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
Support for Ubuntu 25.10 and Ubuntu 26.04 LTS bases
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Charmcraft 4.1 adds experimental support for Ubuntu 25.10 and Ubuntu 26.04 bases.
As an interim base, the behaviour of Charmcraft when packing an Ubuntu 25.10 charm is

> not guaranteed to be stable, but is instead consistent with the evolving changes

Reviewer: > consistent with the evolving changes

Might come across as obtuse. It could be clearer to say it's main purpose is for testing software and charms in preparation for 26.04.

needed for Ubuntu 26.04. See our `Interim bases policy
<https://github.com/canonical/charmcraft/issues/1821>`_ for further information.

> 12-factor app charm HTTP proxy integration

Author: Technically it can integrate with any charm that properly implements the provider side of the `http_proxy` relation. In practice, the only charm I know of that does this is the squid one.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Charms that use any of the 12-factor :ref:`extensions` can now integrate by default with

Reviewer: "by default" doesn't seem to apply here, if they need to set a relation. Maybe:

> extensions now have out-of-the-box support for the Squid Forward Proxy charm. You can connect your charm to it with the `http-proxy` relation.

"connect" is iffy...

the `Squid Forward Proxy <https://charmhub.io/squid-forward-proxy>`_ charm using the
``http-proxy`` relation.

Spring Boot App Profiles

> ~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
Support for Spring Boot profiles
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
